packages/openapi-fetch/src/index.ts

```diff
@@ -89,38 +93,55 @@ export default function createClient<Paths extends {}>(clientOptions: ClientOpti
     return { error, response: response as any };
   }
 
+  type GetPaths = PathsWithMethod<Paths, "get">;
+  type PutPaths = PathsWithMethod<Paths, "put">;
+  type PostPaths = PathsWithMethod<Paths, "post">;
+  type DeletePaths = PathsWithMethod<Paths, "delete">;
+  type OptionsPaths = PathsWithMethod<Paths, "options">;
+  type HeadPaths = PathsWithMethod<Paths, "head">;
+  type PatchPaths = PathsWithMethod<Paths, "patch">;
+  type TracePaths = PathsWithMethod<Paths, "trace">;
+  type GetFetchOptions<P extends GetPaths> = FetchOptions<FilterKeys<Paths[P], "get">>;
+  type PutFetchOptions<P extends PutPaths> = FetchOptions<FilterKeys<Paths[P], "put">>;
+  type PostFetchOptions<P extends PostPaths> = FetchOptions<FilterKeys<Paths[P], "post">>;
+  type DeleteFetchOptions<P extends DeletePaths> = FetchOptions<FilterKeys<Paths[P], "delete">>;
+  type OptionsFetchOptions<P extends OptionsPaths> = FetchOptions<FilterKeys<Paths[P], "options">>;
+  type HeadFetchOptions<P extends HeadPaths> = FetchOptions<FilterKeys<Paths[P], "head">>;
+  type PatchFetchOptions<P extends PatchPaths> = FetchOptions<FilterKeys<Paths[P], "patch">>;
+  type TraceFetchOptions<P extends TracePaths> = FetchOptions<FilterKeys<Paths[P], "trace">>;
+
   return {
     /** Call a GET endpoint */
-    async GET<P extends PathsWithMethod<Paths, "get">>(url: P, init: FetchOptions<FilterKeys<Paths[P], "get">>) {
-      return coreFetch<P, "get">(url, { ...init, method: "GET" } as any);
+    async GET<P extends GetPaths>(url: P, ...init: HasRequiredKeys<GetFetchOptions<P>> extends never ? [GetFetchOptions<P>?] : [GetFetchOptions<P>]) {
+      return coreFetch<P, "get">(url, { ...init[0], method: "GET" } as any);
```

Comment:
The fix was weird, and not straightforward. Essentially, it involved multiple complicated steps:

1. First, convert the function parameters into an array (`...init`) (though for simplicity, we can do this starting at the 2nd param, rather than having to do it for the full spread)
2. Next, we’ll need to take each method’s individual FetchOptions and introspect whether **it has a single required key.** (we needed to create a helper for this)
3. Lastly, expand the parameter array into a tuple type, marking it either optional (`[T?]`) or not (`[T]`) based on the result of #2

Reply:
Again, all of this is very annoying, very strange TypeScript introspection. But that’s the name of the game here, so it’s all worth it… I guess.